# Gradient Descent and Linear Regression with PyTorch

## This tutorial covers the following topics:

* Introduction to linear regression and gradient descent
* Implementing a linear regression model using PyTorch tensors
* Training a linear regression model using the gradient descent algorithm
* Implementing gradient descent and linear regression using PyTorch built-in

## Introduction to Linear Regression

In this tutorial, we'll discuss one of the foundational algorithms in machine learning: *Linear regression*. We'll create a model that predicts crop yields for apples and oranges (*target variables*) by looking at the average temperature, rainfall, and humidity (*input variables or features*) in a region. Here's the training data:

![linear-regression-training-data](https://i.imgur.com/6Ujttb4.png)

In a linear regression model, each target variable is estimated to be a weighted sum of the input variables, offset by some constant, known as a bias :

```
yield_apple  = w11 * temp + w12 * rainfall + w13 * humidity + b1
yield_orange = w21 * temp + w22 * rainfall + w23 * humidity + b2
```

Visually, it means that the yield of apples is a linear or planar function of temperature, rainfall and humidity:

![linear-regression-graph](https://i.imgur.com/4DJ9f8X.png)

The *learning* part of linear regression is to figure out a set of weights `w11, w12,... w23, b1 & b2` using the training data, to make accurate predictions for new data. The _learned_ weights will be used to predict the yields for apples and oranges in a new region using the average temperature, rainfall, and humidity for that region. 

We'll _train_ our model by adjusting the weights slightly many times to make better predictions, using an optimization technique called *gradient descent*. Let's begin by importing Numpy and PyTorch.

In [4]:
import numpy as np
import torch

## Training data

We can represent the training data using two matrices: `inputs` and `targets`, each with one row per observation, and one column per variable.

In [2]:
# Input (temp, rainfall, humidity)
inputs = np.array([[73, 67, 43], 
                   [91, 88, 64], 
                   [87, 134, 58], 
                   [102, 43, 37], 
                   [69, 96, 70]], dtype='float32')

In [3]:
# Targets (apples, oranges)
targets = np.array([[56, 70], 
                    [81, 101], 
                    [119, 133], 
                    [22, 37], 
                    [103, 119]], dtype='float32')

We've separated the input and target variables because we'll operate on them separately. Also, we've created numpy arrays, because this is typically how you would work with training data: read some CSV files as numpy arrays, do some processing, and then convert them to PyTorch tensors.

Let's convert the arrays to PyTorch tensors.

In [4]:
# Convert inputs and targets to tensors
inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)
print(inputs)
print(targets)

tensor([[ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.]])
tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


## Linear regression model from scratch

The weights and biases (`w11, w12,... w23, b1 & b2`) can also be represented as matrices, initialized as random values. The first row of `w` and the first element of `b` are used to predict the first target variable, i.e., yield of apples, and similarly, the second for oranges.

In [5]:
# Weights and biases
w = torch.randn(2, 3, requires_grad=True)
b = torch.randn(2, requires_grad=True)
print(w)
print(b)

tensor([[ 1.7460,  1.0950,  0.9590],
        [ 1.2069, -0.7129, -0.5631]], requires_grad=True)
tensor([-0.9857, -2.0310], requires_grad=True)


`torch.randn` creates a tensor with the given shape, with elements picked randomly from a [normal distribution](https://en.wikipedia.org/wiki/Normal_distribution) with mean 0 and standard deviation 1.

Our *model* is simply a function that performs a matrix multiplication of the `inputs` and the weights `w` (transposed) and adds the bias `b` (replicated for each observation).

![matrix-mult](https://i.imgur.com/WGXLFvA.png)

We can define the model as follows:

In [6]:
def model(x):
  return x @ w.t() + b

`@` represents matrix multiplication in PyTorch, and the `.t` method returns the transpose of a tensor.

The matrix obtained by passing the input data into the model is a set of predictions for the target variables.

In [7]:
# Generate predictions
preds = model(inputs)
print(preds)

tensor([[241.0785,  14.0954],
        [315.6417,   9.0242],
        [353.2746, -25.2168],
        [259.6789,  69.5813],
        [291.7431, -26.6078]], grad_fn=<AddBackward0>)


In [8]:
print(targets)

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


You can see a big difference between our model's predictions and the actual targets because we've initialized our model with random weights and biases. Obviously, we can't expect a randomly initialized model to *just work*.

## Loss function

Before we improve our model, we need a way to evaluate how well our model is performing. We can compare the model's predictions with the actual targets using the following method:

* Calculate the difference between the two matrices (`preds` and `targets`).
* Square all elements of the difference matrix to remove negative values.
* Calculate the average of the elements in the resulting matrix.

The result is a single number, known as the **mean squared error** (MSE).

In [9]:
# MSE loss
def mse(t1, t2):
  diff = t1 - t2
  return torch.sum(diff * diff) / diff.numel()

torch.sum returns the sum of all the elements in a tensor. The .numel method of a tensor returns the number of elements in a tensor. Let's compute the mean squared error for the current predictions of our model.

In [10]:
# Compute loss
loss = mse(preds, targets)
print(loss)

tensor(29519.1211, grad_fn=<DivBackward0>)


Here’s how we can interpret the result: *On average, each element in the prediction differs from the actual target by the square root of the loss*. And that’s pretty bad, considering the numbers we are trying to predict are themselves in the range 50–200. The result is called the *loss* because it indicates how bad the model is at predicting the target variables. It represents information loss in the model: the lower the loss, the better the model.

## Compute gradients

With PyTorch, we can automatically compute the gradient or derivative of the loss w.r.t. to the weights and biases because they have `requires_grad` set to `True`. We'll see how this is useful in just a moment.

In [11]:
loss.backward()

The gradients are stored in the `.grad` property of the respective tensors. Note that the derivative of the loss w.r.t. the weights matrix is itself a matrix with the same dimensions.

In [12]:
# Gradients for weights
print(w)
print(w.grad)

tensor([[ 1.7460,  1.0950,  0.9590],
        [ 1.2069, -0.7129, -0.5631]], requires_grad=True)
tensor([[18502.3086, 18556.2129, 11713.9014],
        [-6587.8711, -9123.5801, -5290.7939]])


## Adjust weights and biases to reduce the loss

The loss is a [quadratic function](https://en.wikipedia.org/wiki/Quadratic_function) of our weights and biases, and our objective is to find the set of weights where the loss is the lowest. If we plot a graph of the loss w.r.t any individual weight or bias element, it will look like the figure shown below. An important insight from calculus is that the gradient indicates the rate of change of the loss, i.e., the loss function's [slope](https://en.wikipedia.org/wiki/Slope) w.r.t. the weights and biases.

If a gradient element is **positive**:

* **increasing** the weight element's value slightly will **increase** the loss
* **decreasing** the weight element's value slightly will **decrease** the loss

![postive-gradient](https://i.imgur.com/WLzJ4xP.png)

If a gradient element is **negative**:

* **increasing** the weight element's value slightly will **decrease** the loss
* **decreasing** the weight element's value slightly will **increase** the loss

![negative=gradient](https://i.imgur.com/dvG2fxU.png)

The increase or decrease in the loss by changing a weight element is proportional to the gradient of the loss w.r.t. that element. This observation forms the basis of _the gradient descent_ optimization algorithm that we'll use to improve our model (by _descending_ along the _gradient_).

We can subtract from each weight element a small quantity proportional to the derivative of the loss w.r.t. that element to reduce the loss slightly.

In [13]:
w
w.grad

tensor([[18502.3086, 18556.2129, 11713.9014],
        [-6587.8711, -9123.5801, -5290.7939]])

In [14]:
with torch.no_grad():
  w -= w.grad * 1e-5
  b -= b.grad * 1e-5

We multiply the gradients with a very small number (`10^-5` in this case) to ensure that we don't modify the weights by a very large amount. We want to take a small step in the downhill direction of the gradient, not a giant leap. This number is called the *learning rate* of the algorithm. 

We use `torch.no_grad` to indicate to PyTorch that we shouldn't track, calculate, or modify gradients while updating the weights and biases.

In [15]:
# Let's verify that the loss is actually lower
loss = mse(preds, targets)
print(loss)

tensor(29519.1211, grad_fn=<DivBackward0>)


Before we proceed, we reset the gradients to zero by invoking the `.zero_()` method. We need to do this because PyTorch accumulates gradients. Otherwise, the next time we invoke `.backward` on the loss, the new gradient values are added to the existing gradients, which may lead to unexpected results.

In [16]:
w.grad.zero_()
b.grad.zero_()
print(w.grad)
print(b.grad)

tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([0., 0.])


## Train the model using gradient descent

As seen above, we reduce the loss and improve our model using the gradient descent optimization algorithm. Thus, we can _train_ the model using the following steps:

1. Generate predictions

2. Calculate the loss

3. Compute gradients w.r.t the weights and biases

4. Adjust the weights by subtracting a small quantity proportional to the gradient

5. Reset the gradients to zero

Let's implement the above step by step.

In [17]:
# Generate predictions
preds = model(inputs)
print(preds)

tensor([[210.1000,  27.2932],
        [274.9760,  26.4349],
        [305.5161,  -4.1903],
        [228.4911,  82.1825],
        [252.9606,  -9.5992]], grad_fn=<AddBackward0>)


In [18]:
# Calculate the loss
loss = mse(preds, targets)
print(loss)

tensor(20607.2734, grad_fn=<DivBackward0>)


In [19]:
# Compute gradients
loss.backward()
print(w.grad)
print(b.grad)

tensor([[15307.4785, 15132.6172,  9599.2227],
        [-5220.6626, -7641.8525, -4379.1577]])
tensor([178.2088, -67.5758])


Let's update the weights and biases using the gradients computed above.

In [20]:
# Adjust weights & reset gradients
with torch.no_grad():
  w -= w.grad * 1e-5
  b -= b.grad * 1e-5
  w.grad.zero_()
  b.grad.zero_()

Let's take a look at the new weights and biases

In [21]:
print(w)
print(b)

tensor([[ 1.4079,  0.7581,  0.7459],
        [ 1.3249, -0.5452, -0.4664]], requires_grad=True)
tensor([-0.9897, -2.0295], requires_grad=True)


With the new weights and biases, the model should have a lower loss

In [22]:
# Calculae loss
preds = model(inputs)
loss = mse(preds, targets)
print(loss)

tensor(14592.9141, grad_fn=<DivBackward0>)


We have already achieved a significant reduction in the loss merely by adjusting the weights and biases slightly using gradient descent.

## Train for multiple epochs

To reduce the loss further, we can repeat the process of adjusting the weights and biases using the gradients multiple times. Each iteration is called an _epoch_. Let's train the model for 100 epochs.

In [27]:
# Train for 150 epochs
for i in range(150):
  preds = model(inputs)
  loss = mse(preds, targets)
  loss.backward()
  with torch.no_grad():
    w -= w.grad * 1e-5
    b -= b.grad * 1e-5
    w.grad.zero_()
    b.grad.zero_()

Once again, let's verify that the loss is now lower

In [28]:
# Calculate loss
preds = model(inputs)
loss = mse(preds, targets)
print(loss)

tensor(109.2436, grad_fn=<DivBackward0>)


The loss is now much lower than its initial value. Let's look at the model's predictions and compare them with the targets.

In [29]:
# Predictions
preds

tensor([[ 59.1928,  73.7956],
        [ 82.2001,  97.2768],
        [115.4456, 135.0644],
        [ 33.7038,  58.6789],
        [ 94.3447, 100.1912]], grad_fn=<AddBackward0>)

In [30]:
# Targets
targets

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])

## Linear regression using PyTorch built-ins

We've implemented linear regression and gradient descent model using some basic tensor operations. However, since this is a common pattern in deep learning, PyTorch provides several built-in functions and classes to make it easy to create and rain models with just a few lines of code

Let's begin by importing the torch.nn package from PyTorch, which contains utility classes for building neural networks.

In [1]:
import torch.nn as nn

As before, we represent the inputs and targets and matrices

In [5]:
# Input (temp, rainfall, humidity)
inputs = np.array([[73, 67, 43], 
                   [91, 88, 64], 
                   [87, 134, 58], 
                   [102, 43, 37], 
                   [69, 96, 70], 
                   [74, 66, 43], 
                   [91, 87, 65], 
                   [88, 134, 59], 
                   [101, 44, 37], 
                   [68, 96, 71], 
                   [73, 66, 44], 
                   [92, 87, 64], 
                   [87, 135, 57], 
                   [103, 43, 36], 
                   [68, 97, 70]], 
                  dtype='float32')

# Targets (apples, oranges)
targets = np.array([[56, 70], 
                    [81, 101], 
                    [119, 133], 
                    [22, 37], 
                    [103, 119],
                    [57, 69], 
                    [80, 102], 
                    [118, 132], 
                    [21, 38], 
                    [104, 118], 
                    [57, 69], 
                    [82, 100], 
                    [118, 134], 
                    [20, 38], 
                    [102, 120]], 
                   dtype='float32')

inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)

In [6]:
inputs

tensor([[ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.],
        [ 74.,  66.,  43.],
        [ 91.,  87.,  65.],
        [ 88., 134.,  59.],
        [101.,  44.,  37.],
        [ 68.,  96.,  71.],
        [ 73.,  66.,  44.],
        [ 92.,  87.,  64.],
        [ 87., 135.,  57.],
        [103.,  43.,  36.],
        [ 68.,  97.,  70.]])

We are using 15 training examples to illustrate how to work with large datasets in small batches

# Dataset and DataLoader

we'll create a TensorDataset, which allows access to rows from inputs and targets as tuples, and provides standard APIs for working with many different types of datasets in PyTorch.

In [7]:
from torch.utils.data import TensorDataset

In [8]:
# Define dataset
train_ds = TensorDataset(inputs, targets)
train_ds[0:5]

(tensor([[ 73.,  67.,  43.],
         [ 91.,  88.,  64.],
         [ 87., 134.,  58.],
         [102.,  43.,  37.],
         [ 69.,  96.,  70.]]), tensor([[ 56.,  70.],
         [ 81., 101.],
         [119., 133.],
         [ 22.,  37.],
         [103., 119.]]))

The TensorDataset allows us to access a small section of the training data using the array indexing notation ([0:5] in the above code). It returns a tuple with two elements. The first element contains the input variables for the selected rows, and the second contains the targets.

We'll also create a DataLoader, which can split the data into batches of a predefined size while training. It also provides other utilities like shuffling and random sampling of the data.

In [9]:
from torch.utils.data import DataLoader

In [10]:
# Define data loader
batch_size = 5
train_dl = DataLoader(train_ds, batch_size, shuffle=True)

We can use the data loader in a for loop,Let's look at an example

In [11]:
for xb, yb in train_dl:
  print(xb)
  print(yb)
  break

tensor([[ 68.,  96.,  71.],
        [102.,  43.,  37.],
        [ 87., 134.,  58.],
        [ 88., 134.,  59.],
        [ 91.,  88.,  64.]])
tensor([[104., 118.],
        [ 22.,  37.],
        [119., 133.],
        [118., 132.],
        [ 81., 101.]])


In each iteration, the data loader returns one batch of data with the given batch size. If shuffle is set to True, it shuffles the training data before creating batches. Shuffling helps randomize the input to the optimization algorithm, leading to a faster reduction in the loss.

## nn.Linear

Instead of initializing the weights and biases manually, we can define the model using the nn.Linear class from PyTorch, which does it automatically.

In [12]:
# Define model
model = nn.Linear(3,2)
print(model.weight)
print(model.bias)

Parameter containing:
tensor([[-0.5702,  0.4501,  0.5295],
        [ 0.5218, -0.0436,  0.0551]], requires_grad=True)
Parameter containing:
tensor([-0.0926, -0.3606], requires_grad=True)


We can use the. odel to generate predictions in the same way as before

In [13]:
# Generate predictions
preds = model(inputs)
preds

tensor([[ 11.2058,  37.1802],
        [ 21.5128,  46.8148],
        [ 41.3211,  42.3908],
        [-19.3090,  53.0283],
        [ 40.8346,  35.3171],
        [ 10.1855,  37.7457],
        [ 21.5921,  46.9135],
        [ 41.2803,  42.9678],
        [-18.2887,  52.4629],
        [ 41.9343,  34.8504],
        [ 11.2851,  37.2790],
        [ 20.4925,  47.3802],
        [ 41.2417,  42.2921],
        [-20.4086,  53.4950],
        [ 41.8549,  34.7517]], grad_fn=<AddmmBackward0>)

## Loss Function
Instead of defining a loss function manually, we can use the built-in loss function mse_loss.

In [14]:
# Import nn.functional
import torch.nn.functional as F

The nn.functional package contains many useful loss functions and several other utilities.

In [15]:
# Define loss function
loss_fn = F.mse_loss

Lets's compute the loss for the current predictions of our model.

In [16]:
loss = loss_fn(model(inputs), targets)
print(loss)

tensor(3642.8992, grad_fn=<MseLossBackward0>)


## Optimizer

Instead of manually manipulating the model's weight and biases using gradients, we can use the optimizer optim.SGD .SGD is short for 'stochastic gradient descent'. The term stochastic indicates that samples are selected in random batches instead of as a single group.


In [17]:
# Define optimizer

In [18]:
opt = torch.optim.SGD(model.parameters(), lr=1e-5)

Note that model.parameters() is passed as an argument to optim.SGD so that the optimizer knows which matrices should be modified during the update step. Also, we can specify a learning rate that controls the amount by which the parameters are modified.

## Train the Model

We are now ready to train the model. We'll follow the same process to implement gradient descent:

1. Generate predictions

2. Calculate the loss

3. Compute gradients w.r.t the weights and biases

4. Adjust the weights by subtracting a small quantity proportional to the gradient

5. Reset the gradients to zero

The only change is that we'll work batches of data instead of processing the entire training data in every iteration. Let's define a utility function fit that trains the model for a given number of epochs.

In [19]:
# Utility function to train the model
def fit(num_epochs, model, loss_fn, opt, train_dl):

  # Repeat for given number of epochs
  for epoch in range(num_epochs):

    # Train with batches of data
    for xb, yb in train_dl:

      # 1. Generate predictions
      pred = model(xb)

      # 2. Calculate loss
      loss = loss_fn(pred, yb)

      # 3. Compute gradients
      loss.backward()

      # 4. Update parameter using gradients
      opt.step()

      # 5. Reset the gradients to zero
      opt.zero_grad()

    # Print the progress
    if (epoch+1) % 10 == 0:
      print('Epoch [{}/{}], Loss: {:.4f}'.format(epoch+1, num_epochs, loss.item()))

Some things to note above:

- We use the data loader defined earlier to get batches of data for every iteration.

- Instead of updating parameters (weights and biases) manually, we use opt.step to perform the update and opt.zero_grad to reset the gradients to zero.

- We've also added a log statement that prints the loss from the last batch of data for every 10th epoch to track training progress. loss.item returns the actual value stored in the loss tensor.

Let's train the model for 100 epochs.

In [20]:
fit(100, model, loss_fn, opt, train_dl)

Epoch [10/100], Loss: 650.8629
Epoch [20/100], Loss: 229.0394
Epoch [30/100], Loss: 78.4432
Epoch [40/100], Loss: 73.3809
Epoch [50/100], Loss: 107.9856
Epoch [60/100], Loss: 44.9165
Epoch [70/100], Loss: 44.9257
Epoch [80/100], Loss: 49.1264
Epoch [90/100], Loss: 35.4123
Epoch [100/100], Loss: 3.2683


Let's generate predictions using our model and verify that they're close to our targets.

In [22]:
# Generates predictions
preds = model(inputs)
preds

tensor([[ 57.1120,  71.9937],
        [ 82.3207,  98.3461],
        [117.4401, 134.7557],
        [ 21.7325,  47.4288],
        [101.7988, 109.0026],
        [ 55.8979,  71.0947],
        [ 82.2118,  97.9540],
        [117.7621, 135.1568],
        [ 22.9466,  48.3279],
        [102.9041, 109.5096],
        [ 57.0031,  71.6016],
        [ 81.1065,  97.4471],
        [117.5490, 135.1478],
        [ 20.6272,  46.9219],
        [103.0129, 109.9016]], grad_fn=<AddmmBackward0>)

In [23]:
# Compare with targets
targets

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 57.,  69.],
        [ 80., 102.],
        [118., 132.],
        [ 21.,  38.],
        [104., 118.],
        [ 57.,  69.],
        [ 82., 100.],
        [118., 134.],
        [ 20.,  38.],
        [102., 120.]])

Indeed, the predictions are quite close to our targets. We have a trained a reasonably good model to predict crop yields for apples and oranges by looking at the average temperature, rainfall, and humidity in a region. We can use it to make predictions of crop yields for new regions by passing a batch containing a single row of input.

In [25]:
model(torch.tensor([[85, 53, 38.]]))

tensor([[37.3291, 57.2430]], grad_fn=<AddmmBackward0>)

The predicted yield of apples is 37.3 tons per hectare and that of oranges is 57.2 tons per hectare,